### ISSUE TITLE
How to dynamically combine APIs?

### ISSUE THREAD

**Issue by banool** [poem-web/poem#364]
Hey, I've looked through all the code and examples I could see but couldn't figure out how to do this. I pretty much want to be able to do this:
```
pub fn get_api_service(
    context: Arc<Context>,
) -> OpenApiService {
    let mut apis = vec![
        AccountsApi {
            context: context.clone(),
        },
        BasicApi {
            context: context.clone(),
        },
        BlocksApi {
            context: context.clone(),
        },
        EventsApi {
            context: context.clone(),
        },
        IndexApi {
            context: context.clone(),
        },
        StateApi {
            context: context.clone(),
        },
        TransactionsApi { context },
    ];

    #[cfg(feature = "failpoints")]
    apis.push(SetFailpointsApi {
        context: context.clone(),
    });

    if some_condition {
        apis.push(OtherApi{});
    }

    OpenApiService::new(apis, "Aptos Node API", "1.0.0")
        .server("/v1")
}
```

But this isn't possible today, `OpenApiService::new` expects a tuple, which cannot be modified once built. Beyond that, making a vec of these API objects (e.g. `Vec<Box<dyn OpenApi>>`) isn't possible since they're not object safe. Rust also insists I define the return type, so instead of just OpenApiService above, I have to have a tuple of all the API types as the first generic type param) but the return type in this case might be different depending on whether I conditionally add additional endpoints.

tl;dr: How do I make a function that returns an OpenApiService that might have different APIs in it?

**Comment by banool** [comment]
Note: I specifically want to return OpenApiService from a function because I need it in a separate tool to generate the spec, vs where in the main server I need it to then run a server from.

**Comment by sunli829** [comment]
Maybe you can consider returning an `Endpoint`.

```rust
pub fn get_api_service(
    context: Arc<Context>,
) -> impl poem::Endpoint {
    let apis = (AccountsApi , BasicApi , ...);

    if some_condition {
         OpenApiService::new((apis , OtherApi), "Aptos Node API", "1.0.0")
            .server("/v1")
            .into_endpoint()
    } else {
         OpenApiService::new(apis, "Aptos Node API", "1.0.0")
            .server("/v1")
            .into_endpoint()
    }
}
```

**Comment by banool** [comment]
If I return an endpoint I don't think I can call `.spec` though :/

**Comment by sunli829** [comment]
It's ugly, but it solves the problem. 😂

```rust
pub fn get_api_service(
    context: Arc<Context>,
) -> (impl poem::Endpoint, String) {
    let apis = (AccountsApi , BasicApi , ...);

    if some_condition {
         let service = OpenApiService::new((apis , OtherApi), "Aptos Node API", "1.0.0").server("/v1");
         (service, service.spec())
    } else {
         let service = OpenApiService::new(apis, "Aptos Node API", "1.0.0").server("/v1");
         (service, service.spec())
    }
}
```
